File: solutions/day05/solution.clj
(def words (->> (slurp "input.txt")
                (clojure.string/split-lines)))

(defn vowel-count
  [word]
  (let [vowels (set "aeiou")]
    (count (filter vowels word))))

(defn has-double
  [word]
  (not (nil? (re-matches #".*(([a-z])\2{1}).*" word))))

(defn has-bad
  ([word] (has-bad word ["ab" "cd" "pq" "xy"]))
  ([word bad]
    (->> (map #(re-pattern (format ".*%s.*" %)) bad)
         (map #(re-matches % word))
         (every? nil?)
          not)))

(defn nice?
  [word]
  (every? true? [(< 2 (vowel-count word))
                 (has-double word)
                 (not (has-bad word))]))

(defn has-repeat-pair
  [word]
  (not (nil? (re-matches #".*([a-z]{2}).*\1.*" word))))

(defn has-repeat-letter
  [word]
  (not (nil? (re-matches #".*([a-z])[a-z]\1.*" word))))

(defn nicer?
  [word]
  (and (has-repeat-pair word) (has-repeat-letter word)))

(println "part 1:" (count (filter nice? words)))
(println "part 2" (count (filter nicer? words)))

File: solutions/day05/solution.py
import re


with open('input.txt', 'r') as fh:
    lines = [l.strip() for l in fh.readlines()]

vowels = 'aeiou'
bad = ['ab', 'cd', 'pq', 'xy']


def nice(word):
    vowel_count = sum(word.count(v) for v in vowels) >= 3
    double = re.match('.*(([a-z])\\2{1}).*', word) is not None
    no_bads = not any(b in word for b in bad)
    return all([vowel_count, double, no_bads])


print 'part 1', len(filter(nice, lines))


def nice_pt2(word):
    repeat_pair = re.match('.*([a-z]{2}).*\\1.*', word) is not None
    repeat_letter = re.match('.*([a-z])[a-z]\\1.*', word) is not None
    return repeat_pair and repeat_letter


print 'part 2', len(filter(nice_pt2, lines))


